### ISSUE TITLE
Improve Stream Performance by Allowing Customizable Concurrency Limits in WebTransport

### ISSUE THREAD

**Issue by cybersoulK** [w3c/webtransport#544]
When using WebTransport to send multiple streams simultaneously, there appears to be a performance bottleneck. Specifically, sending 10,000 streams concurrently results in a delay of approximately 5 seconds for the streams to arrive, which is far from the expected near-instantaneous delivery.

Proposed Enhancement:
To address this performance issue, it would be beneficial to add a set_max_concurrent_uni_streams function. This would allow developers to customize the maximum number of unidirectional streams that can be opened concurrently, providing them with control over the system's limitations and potentially enhancing performance.

Next Steps:
Implement the set_max_concurrent_uni_streams function in the WebTransport API.

**Comment by martinthomson** [comment]
What software (server, client) is your test using?

We have discussed having some means to signal to the browser to raise concurrency limits for a session.  Servers should be able to scale their concurrency accordingly.

**Comment by martinthomson** [comment]
See #446.

**Comment by cybersoulK** [comment]
@martinthomson 

In my application, the server sends many small streams, each with a single message.
These are sent at the start to initialize the world.

My proposal was to let the client decide the limit as a setting, that should match the application requirements. I personally would set 10.000 streams.

#446 seems to propose a dynamic incremental alghoritm, which would not work for me.

My current hacky transport layer over datagrams works better than uni-streams at the moment. (5 seconds vs 0.1 seconds)

**Comment by martinthomson** [comment]
Yeah, I don't think that dynamic response is necessarily a good plan, especially when you know, ahead of time, what you need.  That's why I think we should have a means to ask the browser to open up its limits.

**Comment by aboba** [comment]
@cybersoulK Do you have a small code sample that can reproduce the issue?  We want to verify that this relates to the maximum stream limits, not other issues.

**Comment by jan-ivar** [comment]
Meeting:
- ```js
   new WebTransport({incomingPoolSize: 1000000});
   ```
- Clarifying question of whether limit is absolute or concurrent: MAX_LIMIT is an absolute value in the protocol, but the browser is able to raise that limit, and can take into account concurrent streams when deciding when to raise the protocol limit.

**Comment by jan-ivar** [comment]
A constructor only lets you set it once though. Does it need to be an attribute so the application can change it?

```js
const wt = new WebTransport({maxConcurrentReceiveStreams: 10000});
console.log(wt.maxConcurrentReceiveStreams); // 10000
wt.maxConcurrentReceiveStreams = 20000;
wt.maxConcurrentReceiveStreams = null;
```

**Comment by jan-ivar** [comment]
I've renamed this issue to focus on server→client in contrast to #446.

**Comment by jan-ivar** [comment]
Meeting:
- API numbers are per session, even in pooled situations (browsers do the math)
- Incoming is more correct than receive streams given how the limits are applied
- Need separate control of uni vs bidi
- ```js
   const wt = new WebTransport({
     maxConcurrentIncomingUnidirectionalStreams: 10000,
     maxConcurrentIncomingBidirectionalStreams: 10000,
   });
   ```
- This is a hint to the user agent which may clamp max values

**Comment by jan-ivar** [comment]
Regarding this being a "hint", I'd rather specify a bit more detail to aide WPT testing of this. I think a good precedent to follow here is [jitterBufferTarget](https://w3c.github.io/webrtc-extensions/#rtcrtpreceiver-jitterbuffertarget-rtcrtpreceiver-interface), which means we should probably define a minimum and maximum range that we throw on.

How about [100, 100000]?

**Comment by martinthomson** [comment]
I would think that a minimum of 0 is fine, if potentially not very useful.  The maximum should probably be implementation-defined, but setting a minimum value that an implementation is required to support is worthwhile.

Allowing 100k is a pretty big commitment.  Streams take up more space than jitter buffer samples.  And they will be kept outside of the space that we attribute to the site.  I'd prefer if it were smaller, but would be OK if we ended up with 100k.

**Comment by wilaw** [comment]
[Responses](https://notes.ietf.org/notes-ietf-118-webtrans) received from IETF #118. 

Marten Seemann: Making it configurable seems reasonable, is there any API to see what the max is?

Will: There’s no max API, would you like one?

Marten: Maybe :)

Jan-Ivar: Most of here is on server side, so wouldn’t impact what you’d be able to create client side. There’s a slight difference whether it came from server or client and what direction it is.

Will: Inconclusive feedback, but we appreciate the time.

**Comment by jan-ivar** [comment]
> Allowing 100k is a pretty big commitment.

For sure. I didn't mean to imply a commitment, only a max insane level we all agree to throw on in the API. The user agent would still be allowed to clamp input to its own (lower) maximum. At least that's what [jitterBufferTarget](https://w3c.github.io/webrtc-extensions/#rtcrtpreceiver-jitterbuffertarget-rtcrtpreceiver-interface) does.

**Comment by wilaw** [comment]
"max" implies rejection above a certain limit. This is intended as more of a general hint to the UA. For bikeshedding, how about:

const wt = new WebTransport(url, {
   expectedConcurrentIncomingUnidirectionalStreams: 10000,
   expectedConcurrentIncomingBidirectionalStreams: 30
 });

A valid range for testing might be [0 .. 20000]
